File: library/metaoutput.client/extension/AnyPreview.cpp

#include "../All.hpp"

// extension::AnyPreview ######################################################
void extension::AnyPreview::Connect(MP_STRING application, MP_STRING extension)
{
    try
    {
        {
            extension::AnyTool::Connect(application, extension);
        }
        {
            SetState(STATE::WORK::WAIT);
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
}

void extension::AnyPreview::Disconnect()
{
    try
    {
        if ((s_Items != nullptr) && (MP_VECTOR_EMPTY(s_Items) == false))
        {
            auto a_Size = MP_VECTOR_SIZE_GET(s_Items);
            for (auto i = 0; i < a_Size; i++)
            {
                auto a_Context = MP_VECTOR_GET(s_Items, 0);
                {
                    MP_VECTOR_DELETE(s_Items, 0);
                }
                {
                    extension::AnyTool::Execute(a_Context->m_Name, pattern::data::Message::CONSTANT::PIPE::TERMINATE_REQUEST);
                }
            }
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
}

bool extension::AnyPreview::Register(MP_STRING extension, MP_PTR(AnyPreview) context)
{
    try
    {
        if (s_Items == nullptr)
        {
            s_Items = MP_NEW MP_VECTOR(AnyPreview);
        }
        if ((s_Items != nullptr) && (context != nullptr) && (MP_STRING_EMPTY(extension) == false))
        {
            auto a_Name = GetExtension(extension);
            {
                context->m_Name = a_Name;
                context->m_Thread = nullptr;
            }
            for (auto i = MP_VECTOR_SIZE_GET(s_Items) - 1; i >= 0; i--)
            {
                if (s_Items[i]->m_Name == a_Name)
                {
                    return false;
                }
            }
            {
                MP_THREAD_INITIALIZE(context->m_Thread, __ThreadExecute);
                MP_THREAD_NAME_SET(context->m_Thread, a_Name);
                MP_THREAD_APARTMENT_SET(context->m_Thread, MP_THREAD_APARTMENT_STA);
                MP_THREAD_START(context->m_Thread, context);
            }
            {
                MP_VECTOR_APPEND(s_Items, context);
            }
            return true;
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
    return false;
}

bool extension::AnyPreview::Unregister(MP_STRING extension)
{
    try
    {
        if ((s_Items != nullptr) && (MP_STRING_EMPTY(extension) == false))
        {
            auto a_Name = GetExtension(extension);
            for (auto i = MP_VECTOR_SIZE_GET(s_Items) - 1; i >= 0; i--)
            {
                if (s_Items[i]->m_Name == a_Name)
                {
                    MP_VECTOR_DELETE(s_Items, i);
                    return true;
                }
            }
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
    return false;
}

bool extension::AnyPreview::Execute(MP_STRING url)
{
    try
    {
        auto a_Context1 = (MP_PTR(MP_PIPE_CLIENT))nullptr;
        auto a_Context2 = (MP_PTR(MP_PIPE_WRITESTREAM))nullptr;
        {
            MP_PIPE_CLIENT_INITIALIZE(a_Context1, GetPipeName(GetExtension(url)), MP_PIPE_DIRECTION_OUT);
            MP_PIPE_CLIENT_CONNECT(a_Context1, pattern::data::Message::CONSTANT::PIPE::TIMEOUT);
            MP_PIPE_WRITESTREAM_INITIALIZE(a_Context2, a_Context1);
        }
        if (MP_PIPE_CLIENT_CONNECTED(a_Context1))
        {
            MP_PIPE_WRITESTREAM_WRITE(a_Context2, url);
        }
        {
            MP_PIPE_WRITESTREAM_FINALIZE(a_Context2);
            MP_PIPE_CLIENT_FINALIZE(a_Context1);
        }
        return true;
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
    return false;
}

// Public ##############
MP_STRING extension::AnyPreview::GetExtension(MP_STRING url)
{
    if (MP_STRING_EMPTY(url))
    {
        return "";
    }
    else
    {
        auto a_Index = MP_STRING_LASTINDEXOF(url, ".");
        if (a_Index > 0)
        {
            return MP_STRING_LOWER(MP_STRING_SUBSTRINGEND(url, a_Index));
        }
    }
    return MP_STRING_LOWER(url);
}

MP_STRING extension::AnyPreview::GetFinalText(MP_STRING value)
{
    if (MP_STRING_EMPTY(value) == false)
    {
        auto a_TextBuffer = MP_STRING_BUFFER_GET(value);
        auto a_TextSize = MP_STRING_SIZE_GET(value);
        auto a_Result = MP_NEW System::Text::StringBuilder(a_TextSize);
        auto a_IsFound = false;
        for (auto i = 0; i < a_TextSize; i++)
        {
            switch (a_TextBuffer[i])
            {
            case ' ':
            case '\t':
            case '\r':
            case '\n':
                if (a_IsFound == false)
                {
                    a_Result->Append(" ");
                    a_IsFound = true;
                }
                break;
            default:
                if (a_TextBuffer[i] > 20)
                {
                    a_Result->Append(a_TextBuffer[i]);
                    a_IsFound = false;
                }
                break;
            }
        }
        return MP_STRING_TRIM(a_Result->ToString());
    }
    return "";
}

MP_STRING extension::AnyPreview::GetPipeName(MP_STRING name)
{
    auto a_Context = ":" + MP_CONVERT_STRING_FROM_INT(MP_PROCESS_ID_GET(MP_PROCESS_CURRENT_GET()), 0);
    if (MP_STRING_CONTAINS(name, a_Context))
    {
        return "urn:metaoutput:preview:" + MP_STRING_LOWER(name);
    }
    else
    {
        return "urn:metaoutput:preview:" + MP_STRING_LOWER(name) + a_Context;
    }
}

MP_STRING extension::AnyPreview::GetProperty(MP_STRING name)
{
    auto a_Result = (MP_STRING)"";
    if (MP_STRING_EMPTY(name) == false)
    {
        auto a_Name = "Software\\MetaPlatform\\" + MP_STRING_REPLACE(name, "/", "\\");
        try
        {
            auto a_Index = MP_STRING_LASTINDEXOF(a_Name, "\\");
            if (a_Index > 0)
            {
                MP_REGISTRY a_Context;
                {
                    MP_REGISTRY_INITIALIZE(a_Context, MP_REGISTRY_ROOT_CURRENT_USER, MP_STRING_SUBSTRING(a_Name, 0, a_Index), false);
                }
                {
                    a_Result = MP_REGISTRY_GET(a_Context, MP_STRING_SUBSTRING(a_Name, a_Index + 1, MP_STRING_SIZE_GET(a_Name) - a_Index - 1), "");
                }
                {
                    MP_REGISTRY_FINALIZE(a_Context);
                }
            }
        }
        catch (MP_PTR(MP_EXCEPTION) ex)
        {
            (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
        }
    }
    return a_Result;
}

MP_INT extension::AnyPreview::GetProperty(MP_STRING name, bool isVerified)
{
    if (MP_STRING_EMPTY(name) == false)
    {
        auto a_Result = 0;
        auto a_Context = extension::AnyPreview::GetProperty(name);
        if (MP_STRING_EMPTY(a_Context) == false)
        {
            try
            {
                a_Result = MP_CONVERT_STRING_TO_INT(a_Context);
            }
            catch (MP_PTR(MP_EXCEPTION))
            {
                // This exception can be ignored
            }
        }
        if (isVerified)
        {
            if (name == extension::AnyPreview::VARIABLE::PREVIEW::DOCUMENT_SIZE)
            {
                return MP_MAX(a_Result, 2);
            }
            if (name == extension::AnyPreview::VARIABLE::PREVIEW::BROWSER_SIZE)
            {
                return MP_MAX(a_Result, 10);
            }
            if (name == extension::AnyPreview::VARIABLE::PREVIEW::MEDIA_SIZE)
            {
                return MP_MAX(a_Result, 2);
            }
            if (name == extension::AnyPreview::VARIABLE::PREVIEW::TABLE_SIZE)
            {
                return MP_MAX(a_Result, 100);
            }
        }
        return a_Result;
    }
    return 0;
}

MP_STRING extension::AnyPreview::GetState()
{
    return extension::AnyPreview::GetProperty("METAOUTPUT/PREVIEW/STATE");
}

void extension::AnyPreview::SetState(MP_STRING value)
{
    MP_REGISTRY a_Context;
    try
    {
        MP_REGISTRY_INITIALIZE(a_Context, MP_REGISTRY_ROOT_CURRENT_USER, "Software\\MetaPlatform\\METAOUTPUT\\PREVIEW", false);
        MP_REGISTRY_SET(a_Context, "STATE", value);
        MP_REGISTRY_FINALIZE(a_Context);
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        (MP_NEW atom::Trace())->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
    }
}

// Protected ###########
void extension::AnyPreview::_Execute(MP_PTR(atom::Trace), MP_INT, MP_STRING, MP_STRING)
{
}

// Private #############
void extension::AnyPreview::__Execute(MP_PTR(AnyPreview) sender, MP_PTR(atom::Trace) trace, MP_INT level, MP_STRING url)
{
    try
    {
        auto a_Context = atom::Trace::GetUrlFinal(url);
        {
            SetState(STATE::WORK::EXECUTE);
        }
        try
        {
            if ((MP_FILE_FOUND(a_Context) == false) && (MP_FOLDER_FOUND(a_Context) == false))
            {
                auto a_Context1 = (MP_PTR(MP_WEB_CLIENT))nullptr;
                {
                    a_Context = atom::Trace::GetUrlTemp(url);
                }
                {
                    MP_WEB_CLIENT_INITIALIZE(a_Context1);
                    MP_WEB_CLIENT_DOWNLOAD_FILE_ASYNC(a_Context1, url, a_Context, __DownloadCompleted);
                    MP_WEB_CLIENT_ONPROGRESS_SET(a_Context1, __DownloadProgress);
                }
                while ((GetState() != STATE::WORK::CANCEL) && MP_WEB_CLIENT_ISWORKING(a_Context1))
                {
                    MP_THREAD_SLEEP(50);
                }
                {
                    MP_WEB_CLIENT_CANCEL_ASYNC(a_Context1);
                    MP_WEB_CLIENT_FINALIZE(a_Context1);
                }
                if (MP_FILE_SIZE_GET(a_Context) == 0)
                {
                    a_Context = atom::Trace::GetUrlFinal(url);
                }
            }
        }
        catch (MP_PTR(MP_EXCEPTION))
        {
            a_Context = atom::Trace::GetUrlFinal(url);
        }
        if (MP_FILE_FOUND(a_Context) || MP_FOLDER_FOUND(a_Context) || MP_STRING_CONTAINS(url, "://") || MP_STRING_CONTAINS(url, ":\\\\"))
        {
            {
                trace->SendPreviewMessage(level, NAME::SOURCE::PREVIEW, NAME::EVENT::INFO, url, STATE::MESSAGE::RESEND | pattern::data::Message::STATE::CLEAR | pattern::data::Message::STATE::EXPAND, 0);
            }
            if ((GetState() != STATE::WORK::CANCEL) && (sender != nullptr))
            {
                sender->_Execute(trace, level, url, a_Context);
            }
            if (GetState() == STATE::WORK::CANCEL)
            {
                trace->SendPreviewMessage(level, NAME::SOURCE::PREVIEW, NAME::EVENT::WARNING, url, STATE::MESSAGE::UPDATE, 0);
                trace->SendTraceMessage(0, NAME::SOURCE::ALERT, NAME::EVENT::WARNING, "[[[PREVIEW]]].[[[FAILED]]]: [[[Execution is terminated]]], [" + url + "]");
            }
        }
        else
        {
            trace->SendTraceMessage(0, NAME::SOURCE::ALERT, NAME::EVENT::ERROR, "[[[PREVIEW]]].[[[FAILED]]]: [[[File not found]]], [" + url + "]");
        }
        {
            SetState(STATE::WORK::WAIT);
        }
    }
    catch (MP_PTR(MP_EXCEPTION) ex)
    {
        {
            SetState(STATE::WORK::WAIT);
        }
        {
            trace->SendTraceMessage(0, NAME::SOURCE::ALERT, NAME::EVENT::ERROR, "[[[PREVIEW]]].[[[FAILED]]]: " + MP_EXCEPTION_MESSAGE_GET(ex));
            trace->SendPreviewMessage(level, NAME::SOURCE::PREVIEW, NAME::EVENT::ERROR, url, STATE::MESSAGE::UPDATE, 0);
        }
    }
}

// Private #############
void extension::AnyPreview::MP_THREAD_CALLBACK_MAIN(__ThreadExecute, sender)
{
    auto a_Context = dynamic_cast<MP_PTR(AnyPreview)>(sender);
    auto a_Trace = MP_NEW atom::Trace();
    auto a_IsFound = a_Context == nullptr;
    while (a_IsFound == false)
    {
        try
        {
            auto a_Context1 = (MP_PTR(MP_THREAD_MUTEX))nullptr;
            auto a_Context2 = (MP_PTR(MP_PIPE_SERVER))nullptr;
            auto a_Context3 = (MP_PTR(MP_PIPE_READSTREAM))nullptr;
            {
                MP_THREAD_MUTEX_INITIALIZE(a_Context1, pattern::data::Message::CONSTANT::OUTPUT::MUTEX, false);
                MP_PIPE_SERVER_INITIALIZE(a_Context2, GetPipeName(a_Context->m_Name), MP_PIPE_DIRECTION_IN, 128);
                MP_PIPE_READSTREAM_INITIALIZE(a_Context3, a_Context2);
            }
            while (a_IsFound == false)
            {
                try
                {
                    {
                        MP_PIPE_SERVER_CONNECT(a_Context2);
                    }
                    if (MP_PIPE_SERVER_CONNECTED(a_Context2))
                    {
                        auto a_Context4 = MP_PIPE_READSTREAM_READ(a_Context3);
                        {
                            MP_PIPE_SERVER_DISCONNECT(a_Context2);
                        }
                        if (a_Context4 == pattern::data::Message::CONSTANT::PIPE::TERMINATE_REQUEST)
                        {
                            a_IsFound = true;
                            break;
                        }
                        try
                        {
                            MP_THREAD_MUTEX_LOCK(a_Context1);
                            {
                                __Execute(a_Context, a_Trace->Clear(), 1, a_Context4);
                            }
                            {
                                a_Trace->CommitTransaction();
                                a_Trace->CancelTransaction();
                            }
                            MP_THREAD_MUTEX_UNLOCK(a_Context1);
                        }
                        catch (MP_PTR(MP_EXCEPTION) ex)
                        {
                            MP_THREAD_MUTEX_UNLOCK(a_Context1);
                            a_Trace->CommitTransaction();
                            a_Trace->CancelTransaction();
                            a_Trace->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
                            MP_THREAD_SLEEP(pattern::data::Message::CONSTANT::PIPE::TIMEOUT);
                            break;
                        }
                    }
                }
                catch (MP_PTR(MP_EXCEPTION) ex)
                {
                    a_Trace->CommitTransaction();
                    a_Trace->CancelTransaction();
                    a_Trace->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
                    MP_THREAD_SLEEP(pattern::data::Message::CONSTANT::PIPE::TIMEOUT);
                    break;
                }
            }
            {
                MP_PIPE_READSTREAM_FINALIZE(a_Context3);
                MP_PIPE_SERVER_FINALIZE(a_Context2);
                MP_THREAD_MUTEX_FINALIZE(a_Context1);
            }
        }
        catch (MP_PTR(MP_EXCEPTION) ex)
        {
            a_Trace->CommitTransaction();
            a_Trace->CancelTransaction();
            a_Trace->SendTraceMessage(0, pattern::data::Source::NAME::DIAGNOSTIC, pattern::data::Event::NAME::EXCEPTION, MP_EXCEPTION_MESSAGE_GET(ex));
            break;
        }
    }
}

void extension::AnyPreview::MP_WEB_CLIENT_CALLBACK_DOWNLOAD_COMPLETED(__DownloadCompleted, sender, params)
{
    MP_UNUSED(sender);
    MP_UNUSED(params);
}

void extension::AnyPreview::MP_WEB_CLIENT_CALLBACK_PROGRESS(__DownloadProgress, sender, params)
{
    auto a_Context = dynamic_cast<MP_PTR(MP_WEB_CLIENT)>(sender);
    if (a_Context != nullptr)
    {
        auto a_Context1 = MP_WEB_CLIENT_CALLBACK_PROGRESS_RECEIVED_PERSENT_GET(params);
        if (a_Context1 < 100)
        {
            (MP_NEW atom::Trace())->SendPreviewMessage(0, NAME::SOURCE::PREVIEW, NAME::EVENT::INFO, MP_WEB_CLIENT_CALLBACK_DOWNLOAD_URL_GET(a_Context), STATE::MESSAGE::UPDATE, a_Context1);
        }
        if ((a_Context1 <= 3) && (MP_WEB_CLIENT_CALLBACK_PROGRESS_TOTAL_SIZE_GET(params) > 10000000))
        {
            SetState(STATE::WORK::CANCEL);
        }
    }
}

File: library/metaoutput.client/extension/AnyPreview.cs
#pragma warning disable CA1707 // Identifiers should not contain underscores

using atom;
using Microsoft.Win32;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Globalization;
using System.IO;
using System.IO.Pipes;
using System.Net;
using System.Text;
using System.Threading;

namespace extension
{
#if RELEASE
    [System.Diagnostics.DebuggerStepThrough]
#endif
    public class AnyPreview : basic.AnyObject
    {
        private string m_Name;
        private Thread m_Thread;
        private static List<AnyPreview> s_Items;

        public new class STATE : basic.AnyObject.STATE
        {
            public class WORK
            {
                public const string WAIT = "";
                public const string EXECUTE = "EXECUTE";
                public const string CANCEL = "CANCEL";
            }
        }

        public new class VARIABLE : basic.AnyObject.VARIABLE
        {
            public class DEBUGGING
            {
                public const string DATA_NESTING = "METAOUTPUT/DEBUGGING/DATA.NESTING";
                public const string DATA_TIMEOUT = "METAOUTPUT/DEBUGGING/DATA.TIMEOUT";
                public const string ITEM_COUNT = "METAOUTPUT/DEBUGGING/ITEM.COUNT";
                public const string SHOW_PRIVATE = "METAOUTPUT/DEBUGGING/SHOW.PRIVATE";
                public const string STACK_SIZE = "METAOUTPUT/DEBUGGING/STACK.SIZE";
                public const string STRING_SIZE = "METAOUTPUT/DEBUGGING/STRING.SIZE";
            }

            public class PREVIEW
            {
                public const string BROWSER_SIZE = "METAOUTPUT/PREVIEW/BROWSER.SIZE";
                public const string DOCUMENT_SIZE = "METAOUTPUT/PREVIEW/DOCUMENT.SIZE";
                public const string MEDIA_SIZE = "METAOUTPUT/PREVIEW/MEDIA.SIZE";
                public const string TABLE_SIZE = "METAOUTPUT/PREVIEW/TABLE.SIZE";
            }
        }

        public static void Connect(string application, string extension)
        {
            try
            {
                AnyTool.Connect(application, extension);
                SetState(STATE.WORK.WAIT);
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
        }

        public static void Disconnect()
        {
            try
            {
                if (s_Items != null)
                {
                    var a_Count = s_Items.Count;
                    for (var i = 0; i < a_Count; i++)
                    {
                        var a_Context = s_Items[0];
                        {
                            s_Items.RemoveAt(0);
                        }
                        {
                            AnyTool.Execute(a_Context.m_Name, CONSTANT.MESSAGE.PIPE.TERMINATE_REQUEST);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
        }

        public static bool Register(string extension, AnyPreview context)
        {
            try
            {
                if (s_Items == null)
                {
                    s_Items = new List<AnyPreview>();
                }
                if (s_Items != null && context != null && !string.IsNullOrEmpty(extension))
                {
                    var a_Name = GetExtension(extension);
                    for (var i = s_Items.Count - 1; i >= 0; i--)
                    {
                        if (s_Items[i].m_Name == a_Name)
                        {
                            return false;
                        }
                    }
                    {
                        context.m_Name = a_Name;
                        context.m_Thread = new Thread(__ThreadExecute);
                        context.m_Thread.Name = context.m_Name;
                        context.m_Thread.TrySetApartmentState(ApartmentState.STA);
                        context.m_Thread.Start(context);
                    }
                    {
                        s_Items.Add(context);
                    }
                    return true;
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
            return false;
        }

        public static bool Unregister(string extension)
        {
            try
            {
                if (s_Items != null && !string.IsNullOrEmpty(extension))
                {
                    var a_Name = GetExtension(extension);
                    for (var i = s_Items.Count - 1; i >= 0; i--)
                    {
                        if (s_Items[i].m_Name == a_Name)
                        {
                            s_Items.RemoveAt(i);
                            return true;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
            return false;
        }

        public static bool Execute(string url)
        {
            try
            {
                var a_Context = new NamedPipeClientStream(".", GetPipeName(GetExtension(url)), PipeDirection.Out);
                {
                    a_Context.Connect(CONSTANT.MESSAGE.PIPE.TIMEOUT);
                }
                {
                    var a_Context1 = new StreamWriter(a_Context);
                    if (a_Context.IsConnected)
                    {
                        a_Context1.Write(url);
                        a_Context1.Flush();
                    }
                    {
                        a_Context1.Close();
                    }
                }
                {
                    a_Context.Close();
                }
                return true;
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
            return false;
        }

        public static string GetExtension(string url) // TODO: Replace it to Path.GetExtension(...);
        {
            if (!string.IsNullOrEmpty(url))
            {
                var a_Index = url.LastIndexOf('.');
                if (a_Index > 0)
                {
                    return url.Substring(a_Index, url.Length - a_Index).ToLower(CultureInfo.CurrentCulture);
                }
                else
                {
                    return url.ToLower(CultureInfo.CurrentCulture);
                }
            }
            return "";
        }

        public static string GetFinalText(string value)
        {
            if (!string.IsNullOrEmpty(value))
            {
                var a_Size = value.Length;
                var a_Result = new StringBuilder(a_Size);
                var a_IsFound = false;
                for (var i = 0; i < a_Size; i++)
                {
                    switch (value[i])
                    {
                        case '\t':
                        case '\n':
                        case '\r':
                        case ' ':
                            if (!a_IsFound)
                            {
                                a_Result.Append(' ');
                                a_IsFound = true;
                            }
                            break;
                        default:
                            if (value[i] > '\u0014')
                            {
                                a_Result.Append(value[i]);
                                a_IsFound = false;
                            }
                            break;
                    }
                }
                return a_Result.ToString().Trim();
            }
            return "";
        }

        public static string GetPipeName(string value)
        {
            var a_Context = ":" + System.Diagnostics.Process.GetCurrentProcess().Id.ToString("d0", CultureInfo.InvariantCulture);
            if (value.Contains(a_Context))
            {
                return "urn:metaoutput:preview:" + value.ToLower(CultureInfo.CurrentCulture);
            }
            else
            {
                return "urn:metaoutput:preview:" + value.ToLower(CultureInfo.CurrentCulture) + a_Context;
            }
        }

        public static int GetProperty(string name, bool isVerified)
        {
            if (!string.IsNullOrEmpty(name))
            {
                var a_Result = 0;
                var a_Context = GetProperty(name);
                if (!string.IsNullOrEmpty(a_Context))
                {
                    try
                    {
                        a_Result = Convert.ToInt32(a_Context, CultureInfo.InvariantCulture);
                    }
                    catch (Exception)
                    {
                    }
                }
                if (isVerified)
                {
                    switch (name)
                    {
                        case VARIABLE.PREVIEW.DOCUMENT_SIZE: return Math.Max(a_Result, 2);
                        case VARIABLE.PREVIEW.BROWSER_SIZE: return Math.Max(a_Result, 10);
                        case VARIABLE.PREVIEW.MEDIA_SIZE: return Math.Max(a_Result, 2);
                        case VARIABLE.PREVIEW.TABLE_SIZE: return Math.Max(a_Result, 100);
                    }
                }
                return a_Result;
            }
            return 0;
        }

        public static string GetProperty(string name)
        {
            if (!string.IsNullOrEmpty(name))
            {
                var a_Name = "Software\\MetaPlatform\\" + name.Replace("/", "\\");
                try
                {
                    var a_Index = a_Name.LastIndexOf('\\');
                    if (a_Index > 0)
                    {
                        var a_Result = "";
                        var a_Context = Registry.CurrentUser.CreateSubKey(a_Name.Substring(0, a_Index));
                        {
                            a_Result = a_Context.GetValue(a_Name.Substring(a_Index + 1, a_Name.Length - a_Index - 1), "").ToString();
                        }
                        {
                            a_Context.Close();
                        }
                        return a_Result;
                    }
                }
                catch (Exception ex)
                {
                    new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
                }
            }
            return "";
        }

        public static string GetState()
        {
            return GetProperty("METAOUTPUT/PREVIEW/STATE");
        }

        public static void SetState(string value)
        {
            try
            {
                var a_Context = Registry.CurrentUser.CreateSubKey("Software\\MetaPlatform\\METAOUTPUT\\PREVIEW");
                {
                    a_Context
                        .SetValue("STATE", value);
                }
                {
                    a_Context
                        .Close();
                }
            }
            catch (Exception ex)
            {
                new Trace().SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
            }
        }

        protected virtual void _Execute(Trace trace, int level, string url, string file)
        {
        }

        private static void __Execute(AnyPreview sender, Trace trace, int level, string url)
        {
            try
            {
                SetState(STATE.WORK.EXECUTE);
                {
                    var a_Context = Trace.GetUrlFinal(url);
                    try
                    {
                        if (!File.Exists(a_Context) && !Directory.Exists(a_Context))
                        {
                            {
                                a_Context = Trace.GetUrlTemp(url);
                            }
                            {
                                var a_Context1 = new WebClient();
                                {
                                    a_Context1.DownloadFileCompleted += __DownloadCompleted;
                                    a_Context1.QueryString.Add("url", url);
                                    a_Context1.DownloadFileAsync(new Uri(url), a_Context);
                                    a_Context1.DownloadProgressChanged += __DownloadProgress;
                                }
                                while (GetState() != STATE.WORK.CANCEL && a_Context1.IsBusy)
                                {
                                    Thread.Sleep(50);
                                }
                                {
                                    a_Context1.CancelAsync();
                                }
                                if (new FileInfo(a_Context).Length == 0)
                                {
                                    a_Context = Trace.GetUrlFinal(url);
                                }
                            }
                        }
                    }
                    catch (Exception)
                    {
                        a_Context = Trace.GetUrlFinal(url);
                    }
                    if (File.Exists(a_Context) || Directory.Exists(a_Context) || url.Contains("://") || url.Contains(":\\\\"))
                    {
                        {
                            trace.SendTracePreview(level, NAME.SOURCE.PREVIEW, NAME.EVENT.INFO, url, STATE.MESSAGE.RESEND | STATE.MESSAGE.CLEAR | STATE.MESSAGE.EXPAND, 100);
                        }
                        if (GetState() != STATE.WORK.CANCEL && sender != null)
                        {
                            sender._Execute(trace, level, url, a_Context);
                        }
                        if (GetState() == STATE.WORK.CANCEL)
                        {
                            trace.SendTracePreview(level, NAME.SOURCE.PREVIEW, NAME.EVENT.WARNING, url, STATE.MESSAGE.UPDATE, 100);
                            trace.SendTraceMessage(0, NAME.SOURCE.ALERT, NAME.EVENT.WARNING, "PREVIEW.FAILED: Execution is terminated, [" + url + "]");
                        }
                    }
                    else
                    {
                        trace.SendTraceMessage(0, NAME.SOURCE.ALERT, NAME.EVENT.ERROR, "PREVIEW.FAILED: File not found, [" + url + "]");
                    }
                }
                SetState(STATE.WORK.WAIT);
            }
            catch (Exception ex)
            {
                SetState(STATE.WORK.WAIT);
                trace.SendTraceMessage(0, NAME.SOURCE.ALERT, NAME.EVENT.ERROR, "PREVIEW.FAILED: " + ex.Message);
                trace.SendTracePreview(level, NAME.SOURCE.PREVIEW, NAME.EVENT.ERROR, url, STATE.MESSAGE.UPDATE, 100);
            }
        }

        private static void __ThreadExecute(object sender)
        {
            var a_Context = sender as AnyPreview;
            var a_Trace = new Trace();
            for (var a_IsFound = a_Context == null; !a_IsFound;)
            {
                try
                {
                    var a_Context1 = new Mutex(false, CONSTANT.MESSAGE.OUTPUT.MUTEX);
                    var a_Context2 = new NamedPipeServerStream(GetPipeName(a_Context.m_Name), PipeDirection.In, 128);
                    var a_Context3 = new StreamReader(a_Context2);
                    while (!a_IsFound)
                    {
                        var a_Context4 = "";
                        try
                        {
                            {
                                a_Context2.WaitForConnection();
                            }
                            if (a_Context2.IsConnected)
                            {
                                {
                                    a_Context4 = a_Context3.ReadToEnd();
                                    a_Context2.Disconnect();
                                }
                                if (a_Context4 == CONSTANT.MESSAGE.PIPE.TERMINATE_REQUEST)
                                {
                                    a_IsFound = true;
                                    break;
                                }
                                try
                                {
                                    a_Context1.WaitOne();
                                    a_Trace.CancelTransaction();
                                    __Execute(a_Context, a_Trace, 1, a_Context4);
                                    a_Trace.CommitTransaction();
                                    a_Context1.ReleaseMutex();
                                }
                                catch (Exception ex)
                                {
                                    a_Context1.ReleaseMutex();
                                    a_Trace.CommitTransaction();
                                    a_Trace.SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
                                    a_Trace.SendTracePreview(0, NAME.SOURCE.PREVIEW, NAME.EVENT.EXCEPTION, a_Context4, STATE.MESSAGE.UPDATE, 100);
                                    Thread.Sleep(CONSTANT.MESSAGE.PIPE.TIMEOUT);
                                    break;
                                }
                            }
                        }
                        catch (Exception ex)
                        {
                            a_Trace.CommitTransaction();
                            a_Trace.SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
                            a_Trace.SendTracePreview(0, NAME.SOURCE.PREVIEW, NAME.EVENT.EXCEPTION, a_Context4, STATE.MESSAGE.UPDATE, 100);
                            Thread.Sleep(CONSTANT.MESSAGE.PIPE.TIMEOUT);
                            break;
                        }
                    }
                    {
                        a_Context3.Close();
                        a_Context2.Close();
                    }
                }
                catch (Exception ex)
                {
                    a_Trace.CommitTransaction();
                    a_Trace.CancelTransaction();
                    a_Trace.SendTraceMessage(0, NAME.SOURCE.DIAGNOSTIC, NAME.EVENT.EXCEPTION, ex.Message);
                    break;
                }
            }
        }

        private static void __DownloadCompleted(object sender, AsyncCompletedEventArgs args)
        {
        }

        private static void __DownloadProgress(object sender, DownloadProgressChangedEventArgs args)
        {
            if (sender is WebClient webClient)
            {
                var a_Context = args.ProgressPercentage;
                if (a_Context < 100)
                {
                    new Trace().SendTracePreview(0, NAME.SOURCE.PREVIEW, NAME.EVENT.INFO, webClient.QueryString["url"], STATE.MESSAGE.UPDATE, a_Context);
                }
                if (a_Context <= 3 && args.TotalBytesToReceive > 10000000)
                {
                    SetState(STATE.WORK.CANCEL);
                }
            }
        }
    }
}
